sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupByKey.java

```diff
@@ -162,6 +164,45 @@ public static void applicableTo(PCollection<?> input) {
       throw new IllegalStateException(
           "GroupByKey must have a valid Window merge function.  " + "Invalid because: " + cause);
     }
+
+    // Validate that the trigger does not finish before garbage collection time
+    if (!triggerIsSafe(windowingStrategy)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Unsafe trigger may lose data, see"
+                  + " https://s.apache.org/finishing-triggers-drop-data: %s",
+              windowingStrategy.getTrigger()));
+    }
+  }
+
+  // Note that Never trigger finishes *at* GC time so it is OK, and
+  // AfterWatermark.fromEndOfWindow() finishes at end-of-window time so it is
+  // OK if there is no allowed lateness.
+  private static boolean triggerIsSafe(WindowingStrategy<?, ?> windowingStrategy) {
+    if (!windowingStrategy.getTrigger().mayFinish()) {
```

Comment:
I wonder if `Trigger.mayFinish()` actually should not simply return `false` for all triggers - that is if we can simply drop all the trigger finishing logic.
Let's suppose that a trigger might finish. Then:
 a) if it finishes before window GC time, then it is possibly data loss
 b) if it finishes at window GC time, then it is a no-op
 c) if it finishes after window GC time, then it has no semantic meaning

That seems that we want to avoid a) and b) and c) can be ignored. Do we have any other reason to keep triggers the ability to finish?

Reply:
I agree exactly with all of your analysis. This pull request will forbid triggers of type a).

The meaning of `Trigger.mayFinish()` is to answer this question "is this a trigger that might finish?"

If the answer to that question is "yes" then it might be a type a) trigger. It depends on allowed lateness. There are a couple of very special triggers where the answer "yes" is OK as long as allowed lateness is zero.

Reply:
I'm not sure I follow. If you agree with my analysis, then the only conclusion is that triggers simply *should not finish*. If we want to avoid type a), type b) is no-op (no reason to finish then) and type c) is impossible - why there should still exist "special triggers" that might finish?

Reply:
Yes, I agree that *top level* triggers should not finish. Subtriggers can finish.

 - Consider `Repeatedly.forever(AfterPane.elementCountAtLeast(7))`: The inner trigger finishes, and the outer repeat starts it over.
 - Consider `AfterEach.inOrder(a, b, c)`: Each of `a`, `b`, `c`, finishes to move to the next one.

So it is fine and necessary for a subtrigger to be able to finish.

The question we have to answer is: what do we do when a user tries to set a finishing trigger to be the top level trigger? The two possibilities discussed are:

 - Reject the pipeline, force the user to choose a safe trigger (this PR)
 - Accept the pipeline, but do not actually drop data. Save the data and emit it in the GC pane as we always do. That is #9942 

How do you know whether or not to reject the pipeline? You need a method like `mayFinish`. Any trigger that may finish before GC time must be rejected.

Reply:
Understood. Cool. :+1: